pkg/gardenlet/controller/backupbucket/reconciler.go

```diff
@@ -0,0 +1,441 @@
+	}
+	bb.Status.LastError = lastError
+
+	return c.Status().Patch(ctx, bb, patch)
+}
+
+func updateBackupBucketStatusSucceeded(ctx context.Context, c client.StatusClient, bb *gardencorev1beta1.BackupBucket, message string, clock clock.Clock) error {
+	patch := client.MergeFrom(bb.DeepCopy())
+
+	bb.Status.LastError = nil
+	bb.Status.LastOperation = &gardencorev1beta1.LastOperation{
+		Type:           v1beta1helper.ComputeOperationType(bb.ObjectMeta, bb.Status.LastOperation),
+		State:          gardencorev1beta1.LastOperationStateSucceeded,
+		Progress:       100,
+		Description:    message,
+		LastUpdateTime: metav1.NewTime(clock.Now()),
+	}
+	bb.Status.ObservedGeneration = bb.Generation
+
+	return c.Status().Patch(ctx, bb, patch)
+}
+
+func (r *Reconciler) emptyExtensionSecret(backupBucketName string) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      generateBackupBucketSecretName(backupBucketName),
+			Namespace: r.GardenNamespace,
+		},
+	}
+}
+
+func generateBackupBucketSecretName(backupBucketName string) string {
+	return fmt.Sprintf("bucket-%s", backupBucketName)
+}
+
+func generateGeneratedBackupBucketSecretName(backupBucketName string) string {
+	return v1beta1constants.SecretPrefixGeneratedBackupBucket + backupBucketName
+}
+
+func (r *Reconciler) deployBackupBucketExtensionSecret(ctx context.Context, backupBucket *gardencorev1beta1.BackupBucket) error {
+	coreSecret, err := kutil.GetSecretByReference(ctx, r.GardenClient, &backupBucket.Spec.SecretRef)
+	if err != nil {
+		return err
+	}
+
+	extensionSecret := r.emptyExtensionSecret(backupBucket.Name)
+	_, err = controllerutils.GetAndCreateOrMergePatch(ctx, r.SeedClient, extensionSecret, func() error {
+		extensionSecret.Data = coreSecret.DeepCopy().Data
+		return nil
+	})
+	return err
+}
+
+// deleteGeneratedBackupBucketSecretInGarden deletes generated secret referred by core BackupBucket resource in garden.
+func (r *Reconciler) deleteGeneratedBackupBucketSecretInGarden(ctx context.Context, log logr.Logger, backupBucket *gardencorev1beta1.BackupBucket) error {
+	if backupBucket.Status.GeneratedSecretRef == nil {
+		return nil
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      backupBucket.Status.GeneratedSecretRef.Name,
+			Namespace: backupBucket.Status.GeneratedSecretRef.Namespace,
+		},
+	}
+
+	err := r.GardenClient.Get(ctx, client.ObjectKeyFromObject(secret), secret)
+	if err == nil {
+		if controllerutil.ContainsFinalizer(secret, finalizerName) {
+			log.Info("Removing finalizer from secret", "secret", client.ObjectKeyFromObject(secret))
+			if err := controllerutils.RemoveFinalizers(ctx, r.GardenClient, secret, finalizerName); err != nil {
+				return fmt.Errorf("failed to remove finalizer from secret: %w", err)
+			}
+		}
+	} else if !apierrors.IsNotFound(err) {
+		return fmt.Errorf("failed to get BackupBucket generated secret '%s/%s': %w", secret.Namespace, secret.Name, err)
+	}
```

Comment:
Why are you starting to first handle the non-error case every now and then in this PR? Can you please first handle the error before continuing with the business logic?

Reply:
Not yet addressed

Reply:
Sorry, this was not done by me, so I missed it. Fixed now.